Migrated command docs to python docstrings.

cseed opened the pull request.

jigold commented on python/pyhail/dataset.py:
```diff
@@ -194,6 +194,30 @@ def annotate_global_table(self, input, root, config=None):
+        Load a file as a global annotation.  Consider ``data/genes.txt`` with the form::
+
+          GENE    PLI     EXAC_LOF_COUNT
+          Gene1   0.12312 2
```
Can you try specifying the code language like this:

.. code: bash

jigold commented on python/pyhail/dataset.py:
```diff
@@ -215,6 +239,30 @@ def annotate_global_table(self, input, root, config=None):
+
+        ``condition`` is in sample context so the following symbols are in scope:
+
+          - ``s: Sample``
```
Move each of these bullet points one space to the left. Then it won't render as a block quote.

jigold commented on python/pyhail/dataset.py:
```diff
@@ -451,7 +499,14 @@ def annotate_variants_vds(self, other, code=None, root=None):
 
     def cache(self):
-        """Cache in memory.  cache is the same as persist("MEMORY_ONLY")."""
+        """Mark this dataset to be cached in memory.  ``cache()`` is the same
```
Change cache to :py:meth:\`~pyhail.VariantDataset.cache\` and persist to :func:\`persist("MEMORY_ONLY") <pyhail.VariantDataset.persist>\`

jigold replied:
Fix formatting here.

jigold commented on python/pyhail/dataset.py:
```diff
@@ -541,9 +596,61 @@ def export_genotypes(self, output, condition, types=None, export_ref=False, expo
+        ``fam_expr`` can be used to set the fields in the FAM file.
+        The following fields can be assigned:
+
+          - ``famID: String``
```
Move bullet points to the left one character.

jigold commented on python/pyhail/dataset.py:
```diff
@@ -541,9 +596,61 @@ def export_genotypes(self, output, condition, types=None, export_ref=False, expo
+        ``fam_expr`` is in sample context only and the following
+        symbols are in scope:
+
+          - ``s: Sample``
```
Move bullet points to left one character.

jigold commented on python/pyhail/dataset.py:
```diff
@@ -541,9 +596,61 @@ def export_genotypes(self, output, condition, types=None, export_ref=False, expo
+
+        except:
+
+          - The order among split mutli-allelic alternatives in the BED
```
Move bullet points to left one character

jigold commented on python/pyhail/dataset.py:
```diff
@@ -553,6 +660,32 @@ def export_plink(self, output, fam_expr = 'id = s.id'):
+        One line per sample will be exported.  As ``export_samples``
+        runs in sample context, the following symbols are in scope:
+
+          - ``s: Sample``
```
Move bullet point to left one space.

jigold commented on python/pyhail/dataset.py:
```diff
@@ -553,6 +660,32 @@ def export_plink(self, output, fam_expr = 'id = s.id'):
+
+        **Notes**
+
+        One line per sample will be exported.  As ``export_samples``
```
Replace export_samples with link to function name.

jigold commented on python/pyhail/dataset.py:
```diff
@@ -541,9 +596,61 @@ def export_genotypes(self, output, condition, types=None, export_ref=False, expo
+
+        The BIM file ID field is set to ``CHR:POS:REF:ALT``.
+
+        The ``export_vcf``, ``split_multi`` and ``export_plink``
```
This doesn't make sense to me.

jigold commented on python/pyhail/dataset.py:
```diff
@@ -673,10 +806,124 @@ def write(self, output, overwrite=False):
+
+        ::
+
+          GT: 1/2
```
See if you can get this to be formatted as bash code.

jigold commented on python/pyhail/dataset.py:
```diff
@@ -673,10 +806,124 @@ def write(self, output, overwrite=False):
+
+        In summary:
+
+          - GT: Set to most likely genotype based on the PLs ignoring the filtered allele(s).
```
Move one space to left.

jigold commented on python/pyhail/dataset.py:
```diff
@@ -673,10 +806,124 @@ def write(self, output, overwrite=False):
+
+        The downcoding algorithm would produce the following::
+
+          GT: 0/1
```
bash code

jigold commented on python/pyhail/dataset.py:
```diff
@@ -673,10 +806,124 @@ def write(self, output, overwrite=False):
+        In summary:
+
+          - GT: Downcode the filtered alleles to reference.
+          - AD: The filtered alleles' columns are eliminated and the value is added to the reference, e.g. filtering alleles 1 and 2 transforms ``25,5,10,20`` to ``40,20``.
```
Move one space to left.

jigold commented on python/pyhail/dataset.py:
```diff
@@ -673,10 +806,124 @@ def write(self, output, overwrite=False):
+        Similarly, the depth for the filtered allele in the AD field
+        is added to that of the reference.  If an allele is filtered,
+        this algorithm acts similarly to
+        :py:meth:`~pyhail.VariantDataset.splitmulti`.
```
splitmulti should be split_multi

jigold commented on python/pyhail/dataset.py:
```diff
@@ -673,10 +806,124 @@ def write(self, output, overwrite=False):
+
+        The following symbols are in scope in ``condition``:
+
+          - ``v: Variant``
```
Move one space to left

jigold commented on python/pyhail/dataset.py:
```diff
@@ -673,10 +806,124 @@ def write(self, output, overwrite=False):
+
+        The following symbols are in scope in ``annotation``:
+
+          - ``v: Variant``
```
Move one space to left

jigold commented on python/pyhail/dataset.py:
```diff
@@ -755,7 +1010,17 @@ def filter_samples_list(self, input, keep=True):
-        """Discard all variants, variant annotations and genotypes."""
+        """Discard all variants, variant annotations and genotypes.  Samples,
+          sample annotations and global annotations are retained.
+          This is the same as ``filter_variants_expr('false')``,
```
:func:\`filter_variants_expr('false') <pyhail.VariantDataset.filter_variants_expr>\`

jigold commented on python/pyhail/dataset.py:
```diff
@@ -934,6 +1199,98 @@ def logreg(self, test, y, covariates=None, root=None):
         """Test each variant for association using the logistic regression
         model.
 
+        **Notes**
```
Where are the examples?

jigold commented on python/pyhail/dataset.py:
```diff
@@ -934,6 +1199,98 @@ def logreg(self, test, y, covariates=None, root=None):
 
+        **Notes**
+
+        The ``logreg`` command performs, for each variant, a
```
Use link to logreg command. :py:meth:\`~pyhail.VariantDataset.logreg\`

jigold commented on python/pyhail/dataset.py:
```diff
@@ -934,6 +1199,98 @@ def logreg(self, test, y, covariates=None, root=None):
+        ``sa.cov.age``, ``sa.cov.isFemale``, and ``sa.cov.PC1``, the
+        command::
+
+          logreg -y sa.pheno.isCase -c sa.cov.age,sa.cov.isFemale,sa.cov.PC1
```
This is the wrong command (bash).

jigold commented on python/pyhail/dataset.py:
```diff
@@ -934,6 +1199,98 @@ def logreg(self, test, y, covariates=None, root=None):
+        function <https://en.wikipedia.org/wiki/Sigmoid_function>`_, the
+        genotype :math:`\mathrm{gt}` is coded as 0 for HomRef, 1 for
+        Het, and 2 for HomVar, and the Boolean covariate
+        :math:`\mathrm{isFemale}` is coded as 1 for true (female) and
```
the variable in the formula is "isMale"

jigold commented on python/pyhail/dataset.py:
```diff
@@ -934,6 +1199,98 @@ def logreg(self, test, y, covariates=None, root=None):
+
+        The following R code fits the (standard) logistic, Firth logistic, and linear regression models to this data, where ``x`` is genotype, ``y`` is phenotype, and ``logistf`` is from the logistf package::
+
+          x <- c(rep(0,1000), rep(1,1000), rep(1,10)
```
See if you can format this as R code.

jigold commented on python/pyhail/dataset.py:
```diff
@@ -934,6 +1199,98 @@ def logreg(self, test, y, covariates=None, root=None):
+
+        The resulting p-values for the genotype coefficient are 0.991, 0.00085, and 0.0016, respectively. The erroneous value 0.991 is due to quasi-complete separation. Moving one of the 10 hets from case to control eliminates this quasi-complete separation; the p-values from R are then 0.0373, 0.0111, and 0.0116, respectively, as expected for a less significant association.
+
+        Phenotype and covariate sample annotations may also be specified using `programmatic expressions <reference.html#HailExpressionLanguage>`_ without identifiers, such as::
```
../reference.html

jigold commented on python/pyhail/dataset.py:
```diff
@@ -934,6 +1199,98 @@ def logreg(self, test, y, covariates=None, root=None):
+
+        Phenotype and covariate sample annotations may also be specified using `programmatic expressions <reference.html#HailExpressionLanguage>`_ without identifiers, such as::
+
+          if (sa.isFemale) sa.cov.age else (2 * sa.cov.age + 10)
```
Put this in single quotes so highlighted like a string.

cseed commented:
Ah, inline doesn't do syntax highlighting, looks like.

cseed commented:
I addressed the comments.  For things we don't want to syntax highlight, I used `.. code-block:: text`.

jigold commented on python/pyhail/dataset.py:
```diff
@@ -194,6 +194,32 @@ def annotate_global_table(self, input, root, config=None):
+
+        creates a new global annotation ``global.genes`` with type::
+
+          global.genes: Array[Struct {
```
Can you make this .. code-block text as well?

jigold commented on python/pyhail/dataset.py:
```diff
@@ -215,6 +241,30 @@ def annotate_global_table(self, input, root, config=None):
+
+        ``condition`` is in sample context so the following symbols are in scope:
+
+        - ``s: Sample``
```
Add this relative link to Sample object

:ref:\`sample\`

Likewise variant is

:ref:\`variant\`

jigold commented on python/pyhail/dataset.py:
```diff
@@ -541,9 +597,64 @@ def export_genotypes(self, output, condition, types=None, export_ref=False, expo
+        ``fam_expr`` is in sample context only and the following
+        symbols are in scope:
+
+        - ``s: Sample``
```
:ref:\`sample\`

jigold commented on python/pyhail/dataset.py:
```diff
@@ -541,9 +597,64 @@ def export_genotypes(self, output, condition, types=None, export_ref=False, expo
+
+        - The order among split mutli-allelic alternatives in the BED
+          file may disagree.
+        - PLINK uses the RSID for the BIM file ID.
```
make RSID --> rsID

jigold commented on python/pyhail/dataset.py:
```diff
@@ -553,6 +664,31 @@ def export_plink(self, output, fam_expr = 'id = s.id'):
+
+        **Notes**
+
+        One line per sample will be exported.  As :py:meth:`pyhail.VariantDataset.export_samples` runs in sample context, the following symbols are in scope:
```
Add a tilde before pyhail. Should read :py:meth:\`~pyhail.VariantDataset.export_samples\`

jigold commented on python/pyhail/dataset.py:
```diff
@@ -553,6 +664,31 @@ def export_plink(self, output, fam_expr = 'id = s.id'):
+
+        One line per sample will be exported.  As :py:meth:`pyhail.VariantDataset.export_samples` runs in sample context, the following symbols are in scope:
+
+        - ``s: Sample``
```
See sample comment above

jigold commented on python/pyhail/dataset.py:
```diff
@@ -673,10 +809,128 @@ def write(self, output, overwrite=False):
+        **Example**
+
+        Remove alternate alleles whose allele count is zero and
+        updates the alternate allele count annotation with the new
```
updates should be update

jigold commented on python/pyhail/dataset.py:
```diff
@@ -673,10 +809,128 @@ def write(self, output, overwrite=False):
+
+        The following symbols are in scope in ``condition``:
+
+        - ``v: Variant``
```
Variant should have relative link

:ref:\`variant\`

jigold commented on python/pyhail/dataset.py:
```diff
@@ -673,10 +809,128 @@ def write(self, output, overwrite=False):
+
+        The following symbols are in scope in ``annotation``:
+
+        - ``v: Variant``
```
same as above for relative link

jigold commented on python/pyhail/dataset.py:
```diff
@@ -930,10 +1199,115 @@ def linreg(self, y, covariates='', root='va.linreg', minac=1, minaf=None):
+
+        >>> (hc.read('data/example.vds')
+        >>>   .annotate_samples_table('data/pheno.tsv', root='sa.pheno',
+        >>>     config=TextTableConfig(impute=True))
```
I think you need to use module name  (pyhail.TextTableConfig)

cseed replied:
TextTableConfig is imported into module `pyhail` and I think in the code snippets we should assume the `hc` has been created and `import * from pyhail`.  Reasonable?

jigold commented on python/pyhail/dataset.py:
```diff
@@ -930,10 +1199,115 @@ def linreg(self, y, covariates='', root='va.linreg', minac=1, minaf=None):
+
+        .. math::
+        
+          \mathrm{Prob}(\mathrm{isCase}) = \mathrm{sigmoid}(\\beta_0 + \\beta_1 \, \mathrm{gt} + \\beta_2 \, \mathrm{age} + \\beta_3 \, \mathrm{isFemale} + \\beta_4 \, \mathrm{PC1} + \\varepsilon), \quad \\varepsilon \sim \mathrm{N}(0, \sigma^2)
```
This isn't rendering properly. Not sure what changed since last commit. Is this a read the docs issue?

jigold commented on python/pyhail/dataset.py:
```diff
@@ -930,10 +1199,115 @@ def linreg(self, y, covariates='', root='va.linreg', minac=1, minaf=None):
+        ===== ======================== ====== =====
+        Test  Annotation               Type   Value
+        ===== ======================== ====== =====
+        Wald  ``va.logreg.wald.beta``  Double fit genotype coefficient, :math:`\hat\\beta_1`
```
math is not rendering properly :(

jigold commented on python/pyhail/dataset.py:
```diff
@@ -1202,15 +1573,15 @@ def split_multi(self, propagate_gq=False, keep_star_alleles=False):
 
-         - **va.wasSplit** (*Boolean*) -- true if this variant was
-           originally multiallelic, otherwise false.
+        - **va.wasSplit** (*Boolean*) -- true if this variant was
```
Are we removing block quotes for all annotations? I'm happy with that, but we need to tell the rest of the group.

cseed replied:
I prefer the block quote if you're OK with it.  I put it back.

cseed commented:
A quick look at the source, it isn't loading mathjax.js.  No idea why yet.

cseed commented:
Great feedback.  Addressed comments, back to you.

In the scope lists, I use the format: `variable (*Type*): description`, where the type is in italics but not a hyperlink, but I put a hyperlink in the description when it seemed appropriate.

```
*:ref:`foo`
```

didn't format the hyperlink.  Also, I don't think we can put hyperlinks in double-back-quote literal/code blocks.

I didn't address the math stuff.  I think we can merge this (and other doc migrations) when it is ready and fix that separately.

jigold commented on python/pyhail/dataset.py:
```diff
@@ -1573,15 +1575,14 @@ def split_multi(self, propagate_gq=False, keep_star_alleles=False):
 
         ``split_multi`` adds the following annotations:
```
Can you make this a link to the function? 

:py:meth:\`~pyhail.VariantDataset.split_multi\`

jigold commented on python/pyhail/dataset.py:
```diff
@@ -1573,15 +1575,14 @@ def split_multi(self, propagate_gq=False, keep_star_alleles=False):
 
```
Make data/export.vcf italics

cseed commented:
Back to you.
